File: android/app/src/main/java/me/pjq/rpicar/CaptureVideoFragment.java
package me.pjq.rpicar;

import android.app.Fragment;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.HttpAuthHandler;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;


public class CaptureVideoFragment extends Fragment {
    WebView webView;
    private static final String username = "pjq";
    private static final String passwd = "pjq";

    public CaptureVideoFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment AboutFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static CaptureVideoFragment newInstance() {
        CaptureVideoFragment fragment = new CaptureVideoFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_capture_video, container, false);

        initView(view);
        return view;
    }

    private void initView(View view) {
        webView = (WebView) view.findViewById(R.id.webview);
        webView.loadUrl(CarControllerApiService.Config.CAPTURE_VIDEO_URL());

        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setSupportZoom(true);
        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setLoadWithOverviewMode(true);
        webView.getSettings().setUseWideViewPort(true);

        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                if (request.getUrl().getEncodedPath().endsWith("avi")) {
                    String url = request.getUrl().toString();
                    url =url.replace("http://", "http://" + username + ":" + passwd + "@");
                    Intent intent = new Intent(Intent.ACTION_VIEW);
                    intent.setDataAndType(Uri.parse(url), "video/*");
                    startActivity(intent);

                    return true;
                } else {
                    return super.shouldOverrideUrlLoading(view, request);
                }
            }

            @Override
            public void onReceivedHttpAuthRequest(WebView view, HttpAuthHandler handler, String host, String realm) {
                handler.proceed(username, passwd);
            }
        });
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        getActivity().setTitle(R.string.capturevideo);
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }
}

File: android/app/src/main/java/me/pjq/rpicar/CaptureVideoFragment.kt
package me.pjq.rpicar

import android.app.Fragment
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.HttpAuthHandler
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import me.pjq.rpicar.databinding.FragmentCaptureVideoBinding
import me.pjq.rpicar.utils.Logger

class CaptureVideoFragment : Fragment(), MainNavigationActivity.OnBackKeyListener {
    val TAG: String = "Monitor"
    private var _binding: FragmentCaptureVideoBinding? = null

    // This property is only valid between onCreateView and
// onDestroyView.
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
//        val view = inflater.inflate(R.layout.fragment_capture_video, container, false)
        _binding = FragmentCaptureVideoBinding.inflate(inflater, container, false)
        val view = binding.root

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        initView()
    }

    private fun initView() {
        var videoUrl = CarControllerApiService.Config.CAPTURE_VIDEO_URL()
        Logger.log(TAG, videoUrl)
        binding.webView.loadUrl(videoUrl)
        binding.webView.settings.javaScriptEnabled = true
        binding.webView.settings.setSupportZoom(true)
        binding.webView.settings.builtInZoomControls = true
        binding.webView.settings.loadWithOverviewMode = true
        binding.webView.settings.useWideViewPort = true

        binding.webView.setWebViewClient(object : WebViewClient() {
            override fun shouldOverrideUrlLoading(
                view: WebView,
                request: WebResourceRequest
            ): Boolean {
                if (request.url.encodedPath!!.endsWith("avi")) {
                    var url = request.url.toString()
                    url = url.replace("http://", "http://$username:$passwd@")
                    val intent = Intent(Intent.ACTION_VIEW)
                    intent.setDataAndType(Uri.parse(url), "video/*")
                    startActivity(intent)

                    return true
                } else if (request.url.encodedPath!!.endsWith("disablejpg")) {
                    var url = request.url.toString()
                    url = url.replace("http://", "http://$username:$passwd@")
                    val intent = Intent(Intent.ACTION_VIEW)
                    intent.setDataAndType(Uri.parse(url), "image/*")
                    startActivity(intent)

                    return true
                } else {
                    return super.shouldOverrideUrlLoading(view, request)
                }
            }

            override fun onReceivedHttpAuthRequest(
                view: WebView,
                handler: HttpAuthHandler,
                host: String,
                realm: String
            ) {
                handler.proceed(username, passwd)
            }
        })
    }


    override fun onAttach(context: Context) {
        super.onAttach(context)
        activity.setTitle(R.string.capturevideo)
    }

    override fun onDetach() {
        super.onDetach()
    }

    override fun onBackKeyDown(): Boolean {
        if (binding.webView.canGoBack()) {
            binding.webView.goBack()
            return true
        }

        return false
    }

    companion object {
        private val username = "pjq"
        private val passwd = "pjq"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment AboutFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(): CaptureVideoFragment {
            val fragment = CaptureVideoFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
